packages/ckeditor5-engine/src/model/history.js

```diff
@@ -39,43 +41,170 @@ export default class History {
 		 * Holds all undone operations.
 		 *
 		 * @private
-		 * @member {Set.<module:engine/model/operation/operation~Operation>} module:engine/model/history~History#_undoneOperations
+		 * @type {Set.<module:engine/model/operation/operation~Operation>}
 		 */
 		this._undoneOperations = new Set();
+
+		/**
+		 * A map that allows retrieving the operations fast based on the given base version.
+		 *
+		 * @private
+		 * @type Map.<Number,Number>
+		 */
+		this._baseVersionToOperationIndex = new Map();
+
+		/**
+		 * The history version.
+		 *
+		 * @private
+		 * @type {Number}
+		 */
+		this._version = 0;
+
+		/**
+		 * The gap pairs kept in the <from,to> format.
+		 *
+		 * Anytime the `history.version` is set to a version larger than `history.version + 1`,
+		 * a new <lastHistoryVersion, newHistoryVersion> entry is added to the map.
+		 *
+		 * @private
+		 * @type Map.<number,number>
+		 */
+		this._gaps = new Map();
 	}
 
 	/**
-	 * Adds an operation to the history.
+	 * The version of the last operation in the history.
+	 *
+	 * The history version is incremented automatically when a new operation is added to the history.
+	 * Setting the version manually should be done only in rare circumstances when there is a planned gap
+	 * between history versions. When doing so, the
+	 *
+	 * @type {Number}
+	 */
+	get version() {
+		return this._version;
+	}
+
+	set version( version ) {
+		// Store a gap if there are some operations already in the history and the
+		// new version does not increment the latest one.
+		if ( this.operations.length && version > this._version + 1 ) {
+			this._gaps.set( this._version, version );
+		}
+
+		this._version = version;
+	}
+
+	/**
+	 * The first history operation.
+	 *
+	 * @readonly
+	 * @type {module:engine/model/operation/operation~Operation|undefined}
+	 */
+	get firstOperation() {
+		return this.operations[ 0 ];
+	}
+
+	/**
+	 * The last history operation.
+	 *
+	 * @readonly
+	 * @type {module:engine/model/operation/operation~Operation|undefined}
+	 */
+	get lastOperation() {
+		return this.operations[ this.operations.length - 1 ];
+	}
+
+	/**
+	 * Adds an operation to the history and increments the history version.
```

Comment:
It is a bit internal method but I'd add that the operation's base version should be equal to the history version or an error is thrown.